contracts/royalty-enforcement/RoyaltyWhitelist.sol

```diff
@@ -0,0 +1,124 @@
+*/
+contract RoyaltyWhitelist is ERC165, AccessControl, IRoyaltyWhitelist {
+    ///     =====       Events       =====
+
+    /// @dev Emitted when a target address is added or removed from the whitelist
+    event AddressWhitelistChanged(address indexed target, bool status);
+
+    /// @dev Emitted when a target bytecode is added or removed from the whitelist
+    event BytecodeWhitelistChanged(bytes32 indexed target, bool status);
+
+    ///     =====   State Variables  =====
+
+    /// @dev Only REGISTRAR_ROLE can invoke white listing registration and removal
+    bytes32 public constant REGISTRAR_ROLE = bytes32("REGISTRAR_ROLE");
+
+    /// @dev EOA codehash (see https://eips.ethereum.org/EIPS/eip-1052)
+    bytes32 constant EOA_CODEHASH = keccak256("");
+
+    /// @dev Mapping of whitelisted addresses
+    mapping(address => bool) private addressWhitelist;
+
+    /// @dev Mapping of whitelisted bytecodes
+    mapping(bytes32 => bool) private bytecodeWhitelist;
+
+    ///     =====   Constructor  =====
+
+    /**
+     * @dev Grants `DEFAULT_ADMIN_ROLE` to the supplied `admin` address
+     */
+    constructor(address admin) {
+        _grantRole(DEFAULT_ADMIN_ROLE, admin);
+    }
+
+    ///     =====   View functions  =====
+
+    /// @dev ERC-165 interface support
+    function supportsInterface(
+        bytes4 interfaceId
+    ) public view virtual override(ERC165, AccessControl) returns (bool) {
+        return
+            interfaceId == type(IRoyaltyWhitelist).interfaceId ||
+            super.supportsInterface(interfaceId);
+    }
+
+    /// @dev Returns true if an address is whitelisted, false otherwise
+    function isAddressWhitelisted(
+        address target
+    ) external view override returns (bool) {
+        if (addressWhitelist[target]) {
+            return true;
+        }
+
+        // Retrieve bytecode at target address
+        bytes32 codeHash;
+        assembly {
+            codeHash := extcodehash(target)
+        }
+
+        return bytecodeWhitelist[codeHash];
+    }
+
+    ///     =====  External functions  =====
+
+    /// @dev Add a target address to whitelist
+    function whitelistAddress(
+        address target
+    ) external onlyRole(REGISTRAR_ROLE) {
+        require(target != address(0), "can't whitelist zero address");
+        require(!addressWhitelist[target], "address already whitelisted");
+        addressWhitelist[target] = true;
+        emit AddressWhitelistChanged(target, true);
+    }
+
+    /// @dev Remove a target address from whitelist
+    function unWhitelistAddress(
+        address target
+    ) external onlyRole(REGISTRAR_ROLE) {
+        require(addressWhitelist[target], "address not whitelisted");
+        delete addressWhitelist[target];
+        emit AddressWhitelistChanged(target, false);
+    }
+
+    /// @dev Add a target bytecode to whitelist
+    function whitelistBytecode(
+        bytes32 target
+    ) external onlyRole(REGISTRAR_ROLE) {
```

Comment:
The naming here is quite confusing: we are implicitly calling bytecode hashes just "bytecode" everywhere and ontop of that `target` suggests an address.

Reply:
Also, we need to be careful whitelisting bytecodes that could be proxies.